File: web-panel/src/lib/api/audit.ts
import { api } from "@/lib/api";

export interface AuditLog {
    id: number;
    action: string;
    actor_id: number;
    actor_name: string;
    entity_type: string;
    entity_id: number;
    old_values: string;
    new_values: string;
    ip_address: string;
    request_id: string;
    source: string;
    created_at: string;
}

export interface AuditListParams {
    action?: string;
    entity_type?: string;
    entity_id?: number;
    actor_id?: number;
    date_from?: string;
    date_to?: string;
    offset?: number;
    limit?: number;
}

export interface AuditListResponse {
    success: boolean;
    data: AuditLog[];
    meta: {
        total: number;
        offset: number;
        limit: number;
    };
}

export const auditApi = {
    list: async (params: AuditListParams = {}): Promise<AuditListResponse> => {
        const searchParams = new URLSearchParams();
        if (params.action) searchParams.set("action", params.action);
        if (params.entity_type) searchParams.set("entity_type", params.entity_type);
        if (params.entity_id) searchParams.set("entity_id", String(params.entity_id));
        if (params.actor_id) searchParams.set("actor_id", String(params.actor_id));
        if (params.date_from) searchParams.set("date_from", params.date_from);
        if (params.date_to) searchParams.set("date_to", params.date_to);
        searchParams.set("offset", String(params.offset || 0));
        searchParams.set("limit", String(params.limit || 20));

        const qs = searchParams.toString();
        return await api.getRaw<AuditListResponse>(`/api/v1/admin/audit?${qs}`);
    },
};

File: internal/audit/domain/audit.go
package domain

import (
	"context"
	"time"
)

// AuditAction constants for all admin operations
type AuditAction string

const (
	// User actions
	AuditUserBan         AuditAction = "user.ban"
	AuditUserUnban       AuditAction = "user.unban"
	AuditUserToggleAdmin AuditAction = "user.toggle_admin"
	AuditUserUpdateNotes AuditAction = "user.update_notes"

	// Subscription actions
	AuditSubExtend     AuditAction = "subscription.extend"
	AuditSubRevoke     AuditAction = "subscription.revoke"
	AuditSubPause      AuditAction = "subscription.pause"
	AuditSubResume     AuditAction = "subscription.resume"
	AuditSubSetData    AuditAction = "subscription.set_data"
	AuditSubAddData    AuditAction = "subscription.add_data"
	AuditSubResetData  AuditAction = "subscription.reset_data"
	AuditSubSetExpiry  AuditAction = "subscription.set_expiry"
	AuditSubRegenKey   AuditAction = "subscription.regen_key"
	AuditSubRename     AuditAction = "subscription.rename"
	AuditSubDelete     AuditAction = "subscription.delete"
	AuditSubBulk       AuditAction = "subscription.bulk"
	AuditSubSetDataLim AuditAction = "subscription.set_data_limit"
	// AuditSubViewAccessHistory: admin requested per-subscription request log.
	// Logged because access-history reveals destinations the user contacted —
	// admins should be accountable for these queries.
	AuditSubViewAccessHistory      AuditAction = "subscription.view_access_history"
	AuditSubSearchAccessHistory    AuditAction = "subscription.search_access_history"
	AuditAccessHistoryGlobalSearch AuditAction = "access_history.global_search"

	// Node actions
	AuditNodeCreate AuditAction = "node.create"
	AuditNodeUpdate AuditAction = "node.update"
	AuditNodeDelete AuditAction = "node.delete"
	AuditNodeWipe   AuditAction = "node.wipe"
	AuditNodeNuke   AuditAction = "node.nuke"

	// Xray binary actions
	AuditXrayBinaryUpload AuditAction = "xray.binary_upload"
	AuditXrayBinaryDelete AuditAction = "xray.binary_delete"

	// Settings actions
	AuditSettingsUpdate AuditAction = "settings.update"
	AuditPasswordChange AuditAction = "auth.password_change"
	AuditAdminLogin     AuditAction = "auth.admin_login"

	// Data management
	AuditDataExport AuditAction = "data.export"
	AuditDataBackup AuditAction = "data.backup"

	// Backup & Restore actions
	AuditBackupDelete    AuditAction = "backup.delete"
	AuditBackupDownload  AuditAction = "backup.download"
	AuditRestoreStart    AuditAction = "restore.start"
	AuditRestoreComplete AuditAction = "restore.complete"
	AuditRestoreFailed   AuditAction = "restore.failed"

	// Database management
	AuditDatabaseCleanup AuditAction = "database.cleanup"

	// Server management
	AuditServerRestart AuditAction = "server.restart"
)

// AuditLog represents an audit log entry
type AuditLog struct {
	ID         uint      `gorm:"primaryKey" json:"id"`
	Action     string    `gorm:"size:100;index" json:"action"`
	ActorID    uint      `gorm:"index" json:"actor_id"`
	ActorName  string    `gorm:"size:255" json:"actor_name"`
	EntityType string    `gorm:"size:50;index" json:"entity_type"`
	EntityID   uint      `gorm:"index" json:"entity_id"`
	OldValues  string    `gorm:"type:text" json:"old_values,omitempty"`
	NewValues  string    `gorm:"type:text" json:"new_values,omitempty"`
	IPAddress  string    `gorm:"size:45" json:"ip_address"`
	RequestID  string    `gorm:"size:36;index" json:"request_id"`
	Source     string    `gorm:"size:20" json:"source"`
	CreatedAt  time.Time `gorm:"index" json:"created_at"`
}

func (AuditLog) TableName() string {
	return "audit_logs"
}

// AuditListFilters holds filtering parameters for listing audit logs
type AuditListFilters struct {
	Action     string
	ActorID    uint
	EntityType string
	EntityID   uint
	DateFrom   *time.Time
	DateTo     *time.Time
	Offset     int
	Limit      int
}

// AuditLogRepository defines the repository interface
type AuditLogRepository interface {
	Create(ctx context.Context, entry *AuditLog) error
	List(ctx context.Context, filters AuditListFilters) ([]*AuditLog, int64, error)
	CleanupOlderThan(ctx context.Context, days int) (int64, error)
}

// AuditLogUsecase defines the usecase interface
type AuditLogUsecase interface {
	Log(ctx context.Context, entry *AuditLog)
	List(ctx context.Context, filters AuditListFilters) ([]*AuditLog, int64, error)
	Cleanup(ctx context.Context, days int) (int64, error)
	Stop()
}
